File: app/src/main/java/com/dpis/module/diagnostics/DpisLogEntry.java
package com.dpis.module.diagnostics;

public final class DpisLogEntry {
    public final long timestampMillis;
    public final String timestamp;
    public final String level;
    public final String source;
    public final String process;
    public final String modulePackage;
    public final String tag;
    public final String message;
    public final boolean external;

    public DpisLogEntry(String timestamp,
            String level,
            String process,
            String modulePackage,
            String tag,
            String message,
            boolean external) {
        this(0L, timestamp, level, external ? "LSPosed" : "DPIS",
                process, modulePackage, tag, message, external);
    }

    public DpisLogEntry(long timestampMillis,
            String timestamp,
            String level,
            String source,
            String process,
            String modulePackage,
            String tag,
            String message,
            boolean external) {
        this.timestampMillis = timestampMillis;
        this.timestamp = timestamp != null ? timestamp : "";
        this.level = level != null ? level : "";
        this.source = source != null ? source : "";
        this.process = process != null ? process : "";
        this.modulePackage = modulePackage != null ? modulePackage : "";
        this.tag = tag != null ? tag : "";
        this.message = message != null ? message : "";
        this.external = external;
    }
}

File: app/src/main/java/com/dpis/module/diagnostics/DpisLogEntry.kt
package com.dpis.module.diagnostics

/** Immutable normalized log entry shared by the parser, exporter, and log page. */
class DpisLogEntry private constructor(
    @JvmField val timestampMillis: Long,
    @JvmField val timestamp: String,
    @JvmField val level: String,
    @JvmField val source: String,
    @JvmField val process: String,
    @JvmField val modulePackage: String,
    @JvmField val tag: String,
    @JvmField val message: String,
    @JvmField val external: Boolean,
    @Suppress("UNUSED_PARAMETER") marker: Unit,
) {
    constructor(
        timestamp: String?,
        level: String?,
        process: String?,
        modulePackage: String?,
        tag: String?,
        message: String?,
        external: Boolean,
    ) : this(
        timestampMillis = 0L,
        timestamp = timestamp.orEmpty(),
        level = level.orEmpty(),
        source = if (external) "LSPosed" else "DPIS",
        process = process.orEmpty(),
        modulePackage = modulePackage.orEmpty(),
        tag = tag.orEmpty(),
        message = message.orEmpty(),
        external = external,
        marker = Unit,
    )

    constructor(
        timestampMillis: Long,
        timestamp: String?,
        level: String?,
        source: String?,
        process: String?,
        modulePackage: String?,
        tag: String?,
        message: String?,
        external: Boolean,
    ) : this(
        timestampMillis = timestampMillis,
        timestamp = timestamp.orEmpty(),
        level = level.orEmpty(),
        source = source.orEmpty(),
        process = process.orEmpty(),
        modulePackage = modulePackage.orEmpty(),
        tag = tag.orEmpty(),
        message = message.orEmpty(),
        external = external,
        marker = Unit,
    )
}
